cpp/src/io/parquet/parquet_reader.cu

```diff
@@ -0,0 +1,819 @@
+    const hostdevice_vector<parquet::gpu::PageInfo> &pages) {
+
+  for (size_t c = 0, page_count = 0; c < chunks.size(); c++) {
+    chunks[c].max_num_pages = chunks[c].num_data_pages + chunks[c].num_dict_pages;
+    chunks[c].page_info = pages.device_ptr(page_count);
+    page_count += chunks[c].max_num_pages;
+  }
+
+  CUDA_TRY(cudaMemcpyAsync(chunks.device_ptr(), chunks.host_ptr(),
+                           chunks.memory_size(), cudaMemcpyHostToDevice));
+  CUDA_TRY(parquet::gpu::DecodePageHeaders(chunks.device_ptr(), chunks.size()));
+  CUDA_TRY(cudaMemcpyAsync(pages.host_ptr(), pages.device_ptr(),
+                           pages.memory_size(), cudaMemcpyDeviceToHost));
+  CUDA_TRY(cudaStreamSynchronize(0));
+
+  LOG_PRINTF("[+] Page Header Information\n");
+  for (size_t i = 0; i < pages.size(); i++) {
+    LOG_PRINTF(
+        " %2zd: comp_size=%d, uncomp_size=%d, num_values=%d, chunk_row=%d, "
+        "num_rows=%d\n     chunk_idx=%d, flags=%d, encoding=%d, def_level=%d "
+        "rep_level=%d, valid_count=%d\n",
+        i, pages[i].compressed_page_size, pages[i].uncompressed_page_size,
+        pages[i].num_values, pages[i].chunk_row, pages[i].num_rows,
+        pages[i].chunk_idx, pages[i].flags, pages[i].encoding,
+        pages[i].definition_level_encoding, pages[i].repetition_level_encoding,
+        pages[i].valid_count);
+  }
+}
+
+/**
+ * @brief Decompresses the page data, at page granularity
+ *
+ * @param[in] chunks List of column chunk descriptors
+ * @param[in] pages List of page information
+ *
+ * @return uint8_t* Device pointer to decompressed page data
+ **/
+uint8_t *decompress_page_data(
+    const hostdevice_vector<parquet::gpu::ColumnChunkDesc> &chunks,
+    const hostdevice_vector<parquet::gpu::PageInfo> &pages) {
+
+  auto for_each_codec_page = [&](parquet::Compression codec,
+                                 const std::function<void(size_t)> &f) {
+    for (size_t c = 0, page_count = 0; c < chunks.size(); c++) {
+      const auto page_stride = chunks[c].max_num_pages;
+      if (chunks[c].codec == codec) {
+        for (int k = 0; k < page_stride; k++) {
+          f(page_count + k);
+        }
+      }
+      page_count += page_stride;
+    }
+  };
+
+  // Brotli scratch memory for decoding
+  rmm::device_vector<uint8_t> debrotli_scratch;
+
+  // Count the exact number of compressed pages
+  size_t num_compressed_pages = 0;
+  size_t total_decompressed_size = 0;
+  std::array<std::pair<parquet::Compression, size_t>, 3> codecs{
+      std::make_pair(parquet::GZIP, 0), std::make_pair(parquet::SNAPPY, 0),
+      std::make_pair(parquet::BROTLI, 0)};
+
+  for (auto &codec : codecs) {
+    for_each_codec_page(codec.first, [&](size_t page) {
+      total_decompressed_size += pages[page].uncompressed_page_size;
+      codec.second++;
+      num_compressed_pages++;
+    });
+    if (codec.first == parquet::BROTLI && codec.second > 0) {
+      debrotli_scratch.resize(get_gpu_debrotli_scratch_size(codec.second > 0));
```

Comment:
I think this was meant to be "get_gpu_debrotli_scratch_size(codec.second)"

Reply:
Fixed, thanks!